Handle different slopes in mmaps

jackpoz opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Handle different slopes in mmaps using different area flags that can be included/excluded when querying mmaps. 
- NAV_AREA_GROUND will be used for ground with slope in the range [0, 55] while NAV_AREA_GROUND_STEEP will be used for ground with slope in the range of (55, 70]. This allows to use the [0, 55] range for Players and idle Creatures and to use the (55, 70] range for Creatures in combat/evading (chasing Players in places usually not reachable)
- NAV_AREA_GROUND_STEEP takes priority over NAV_AREA_GROUND
- Remove --maxAngle parameter from mmaps_generator to keep all parameters hardcoded in MapBuilder::GetMapSpecificConfig() (one is not supposed to try random angles anyway)
- It's required to re-generate mmaps

This is how the tile looks in recastdemo. Light blue is NAV_AREA_GROUND, dark blue is NAV_AREA_GROUND_STEEP:
![image](https://user-images.githubusercontent.com/1153754/83970467-2aa2c700-a8d6-11ea-91a5-80a742beff9a.png)

Example path going up the hill using both area types:
![image](https://user-images.githubusercontent.com/1153754/83970488-3e4e2d80-a8d6-11ea-8caf-2d22f627503f.png)

Same path, this time excluding NAV_AREA_GROUND_STEEP:
![image](https://user-images.githubusercontent.com/1153754/83970504-4d34e000-a8d6-11ea-876a-06de1f06d1e3.png)

https://github.com/jackpoz/recastnavigation/tree/fixes/3.3.5-blink-55-degrees is the recastdemo branch that supports this

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #24644


**Tests performed:** (Does it build, tested in-game, etc.)
None


**Known issues and TODO list:** (add/remove lines as needed)

- [x] Implement the change in mmaps_generator
- [x] Implement the change in worldserver
- [x] Cleanup the code to reduce changes in recastnavigation
- [x] raise the steep angle from 70° to 85°


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

jackpoz committed:
Tools/MMAPs: Remove input argument --maxAngle

Remove input argument --maxAngle . This should be just hardcoded in MapBuilder::GetMapSpecificConfig() so all settings are easily found in the same place instead of being spread around.

jackpoz committed:
Tools/MMAPs: Add new area type NAV_AREA_GROUND_STEEP

Add new area type NAV_AREA_GROUND_STEEP for ground with slope in the range (55, 70] .
NAV_AREA_GROUND is used for ground with range [0, 55] .
NAV_AREA_GROUND_STEEP takes priority over NAV_AREA_GROUND.

jackpoz committed:
Tools/MMAPs: Fix NAV_GROUND_STEEP flag not being saved in the mmtile

illfated commented:
Off-topic/curiosity: Which map are the example images taken from, is it some PvP or raid map? I notice something looking like a Night Elf tower, but I am still struggling too much to recognize it.

jackpoz committed:
Core/PathFinding: Implement NAV_GROUND_STEEP flag

Implement NAV_GROUND_STEEP flag, used only by Creatures that are in combat or evading.

jackpoz commented:
@illfated that's WSG battleground at the location of https://github.com/TrinityCore/TrinityCore/issues/24644

jackpoz commented:
Current status of this PR: it works like a charm but I would like to change area ids to avoid having NAV_AREA_GROUND and RC_WALKABLE_AREA with the same id 63 (that will required to re-generate mmaps yet again but shouldn't affect anything ingame).

jackpoz committed:
Distinguish between RC_WALKABLE_AREA and NAV_AREA_GROUND.

jackpoz committed:
Allow mobs in combat to walk on steeps up to 80°

jackpoz committed:
Allow mobs in combat to walk on steeps up to 85°.
Disable rcFilterLedgeSpans() filter as it removed long steep spans.

Faq commented on src/tools/mmaps_generator/MapBuilder.cpp:
```diff
@@ -1015,7 +1016,7 @@ namespace MMAP
         config.cs = tileConfig.BASE_UNIT_DIM;
         config.ch = tileConfig.BASE_UNIT_DIM;
-        config.walkableSlopeAngle = 80;
+        config.walkableSlopeAngle = 85;
```
just wondering why changed this.

jackpoz replied:
70° and all filters
![image](https://user-images.githubusercontent.com/1153754/84574141-ea938680-ada4-11ea-9f62-b8810f57b345.png)

80° and all filters
![image](https://user-images.githubusercontent.com/1153754/84574159-f5e6b200-ada4-11ea-8ae6-faa4002032fb.png)

80° and exclude ledge span filter
![image](https://user-images.githubusercontent.com/1153754/84574170-fe3eed00-ada4-11ea-9ad5-2c3709df4582.png)

85° and all filters
![image](https://user-images.githubusercontent.com/1153754/84574179-07c85500-ada5-11ea-8c1c-5b0cf91761a2.png)

85° and exclude ledge span filter
![image](https://user-images.githubusercontent.com/1153754/84574190-1151bd00-ada5-11ea-9bb8-2f011383a35b.png)

jackpoz committed:
Increase cost of steep spans to try making creatures walk around obstacles instead of walking on them

jackpoz committed:
Revert last commit

jackpoz commented:
PR and test realm updated, mobs will try a more direct route to reach the target when in combat.
